### PR TITLE
Introduce `CallPath` to `TyConstantDeclaration`

### LINKED ISSUE #4217: Constants in the root module override other constants with the same name in the submodules
## Problem 

Constants in submodules overriden by the root module if they share the same name. To give an example:

main.sw:

```rust
contract;

dep pkga;

const TEST_CONST: u64 = 20;

abi MyContract {
    fn test_function() -> bool;
}

impl MyContract for Contract {
    fn test_function() -> bool {
        true
    }
}

#[test(should_panic)]
fn test_success() {
  assert(TEST_CONST == pkga::TEST_CONST)
}
```

pkga.sw:
```rust
library pkga;

pub const TEST_CONST: u64 = 10;
```

If `forc test` is run with this setup , the test would be failing but it shouldn't as it is a should revert test and `TEST_CONST` had to be different. Logging them shows that both of them is `20`, the value at the root submodule.

## Possible cause

We had a chat with @mitchmindtree yesterday and made a quick investigation in the compiler. It might be possible that problem is arising from `compile_const_decl` function which compiles const expression to a `Value`. It only accepts a constant name and some look-up variables which contains the namespace. The function then searches the namespace only using the name but in the example above although we are explicitly using  the constant from submodule the name is still `TEST_CONST`. This causes the symbol checking in the namespace to found the `TEST_CONST` in the root module (instead of `pkga::TEST_CONST`)

## Possible solution

`compile_const_decl`  should also be accepting the `path`, and module look-up function should accept the path and the name of the variable so that it can differentiate between the symbols that has the same name in root and submodules.

This is blocking me in #4156. Related to #4162

### PR DESCRIPTION
**Pull request by mohammadfawaz**
## Description
Closes #4217 

Main changes: 
* `TyConstantDeclaration` now contains a `CallPath` instead of an `Ident` to identify the path to and the name of the constant declaration.
* Use the full path, converted to a vector of strings, when looking up and adding constants in `const_eval.rs`. This helps identifying constants with the same name more uniquly.
* Amend the function `to_fullpath` to always complete the path even in situations where only some of the prefixes are present. This is useful for when expressions use a delineated path, especially in internal libraries. 

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by kayagokalp — APPROVED** [review]
Thank you very much for this! Tested by merging with #4156 and everything works! This also closes #4162 I opened it as I was thinking to merge #4156 before this fix

**Review by otrho — APPROVED** [review]

File: sway-lsp/src/traverse/typed_tree.rs [hunk 1]
```diff
@@ -519,16 +519,10 @@ impl<'a> TypedTree<'a> {
                     self.collect_call_path_prefixes(&call_path.prefixes);
                 }
 
-                let span = if let Some(call_path) = call_path {
```
Comment (JoshuaBatty) [inline sway-lsp/src/traverse/typed_tree.rs:522]: Is there any reason to remove this check? We want to make sure if the call path looks something like `foo::bar::const` that the span only contains the `const` token and not the `foo::bar` span info.

Comment (mohammadfawaz) [reply sway-lsp/src/traverse/typed_tree.rs:522]: I see your point. What I did is update the span of VariableExpression to always point to the name of the const in the expression even if a call path is provided, so the check is no longer necessary here. But I'm now wondering it this is the right approach. The span of the suffix of the call paths may point to the const declaration after my changes. I think I can do better here though.

Comment (mohammadfawaz) [reply sway-lsp/src/traverse/typed_tree.rs:522]: Merging for now and will deal with the span thing tonight or tomorrow.

Comment (JoshuaBatty) [reply sway-lsp/src/traverse/typed_tree.rs:522]: Ok great, happy for this to be addressed in a follow up PR so @kayagokalp can be unblocked.

**Pull request merged**